File: Same/services/implementations/EventService.cs
using Microsoft.EntityFrameworkCore;
using Same.Data;
using Same.Models.DTOs.Responses;
using Same.Services.Interfaces;

namespace Same.Services.Implementations
{
    public class EventService : IEventService
    {
        private readonly ApplicationDbContext _context;

        public EventService(ApplicationDbContext context)
        {
            _context = context;
        }

        public Task<ApiResponse<EventResponse>> CreateEventAsync(Guid creatorId, CreateEventRequest request)
        {
            return Task.FromResult(ApiResponse<EventResponse>.ErrorResult("Event service not fully implemented yet"));
        }

        public Task<ApiResponse<EventResponse>> GetEventByIdAsync(Guid eventId, Guid? currentUserId = null)
        {
            return Task.FromResult(ApiResponse<EventResponse>.ErrorResult("Event service not fully implemented yet"));
        }

        public Task<ApiResponse<List<EventResponse>>> GetEventsByCreatorAsync(Guid creatorId)
        {
            return Task.FromResult(ApiResponse<List<EventResponse>>.ErrorResult("Event service not fully implemented yet"));
        }

        public Task<ApiResponse<List<EventResponse>>> SearchEventsAsync(SearchEventsRequest request, Guid? currentUserId = null)
        {
            return Task.FromResult(ApiResponse<List<EventResponse>>.ErrorResult("Event service not fully implemented yet"));
        }

        public Task<ApiResponse<EventResponse>> UpdateEventAsync(Guid eventId, Guid creatorId, UpdateEventRequest request)
        {
            return Task.FromResult(ApiResponse<EventResponse>.ErrorResult("Event service not fully implemented yet"));
        }

        public Task<ApiResponse<bool>> DeleteEventAsync(Guid eventId, Guid creatorId)
        {
            return Task.FromResult(ApiResponse<bool>.ErrorResult("Event service not fully implemented yet"));
        }

        public Task<ApiResponse<bool>> JoinEventAsync(Guid eventId, Guid userId)
        {
            return Task.FromResult(ApiResponse<bool>.ErrorResult("Event service not fully implemented yet"));
        }

        public Task<ApiResponse<bool>> LeaveEventAsync(Guid eventId, Guid userId)
        {
            return Task.FromResult(ApiResponse<bool>.ErrorResult("Event service not fully implemented yet"));
        }

        public Task<ApiResponse<List<UserResponse>>> GetEventParticipantsAsync(Guid eventId)
        {
            return Task.FromResult(ApiResponse<List<UserResponse>>.ErrorResult("Event service not fully implemented yet"));
        }

        public Task<ApiResponse<List<EventResponse>>> GetUserEventsAsync(Guid userId)
        {
            return Task.FromResult(ApiResponse<List<EventResponse>>.ErrorResult("Event service not fully implemented yet"));
        }

        public Task<ApiResponse<bool>> InviteToEventAsync(Guid eventId, Guid creatorId, List<Guid> userIds)
        {
            return Task.FromResult(ApiResponse<bool>.ErrorResult("Event service not fully implemented yet"));
        }

        public Task<ApiResponse<List<EventResponse>>> GetNearbyEventsAsync(decimal latitude, decimal longitude, double radiusKm = 10, Guid? currentUserId = null)
        {
            return Task.FromResult(ApiResponse<List<EventResponse>>.ErrorResult("Event service not fully implemented yet"));
        }

        public Task<ApiResponse<List<EventResponse>>> GetEventsByHobbyAsync(Guid hobbyId, Guid? currentUserId = null)
        {
            return Task.FromResult(ApiResponse<List<EventResponse>>.ErrorResult("Event service not fully implemented yet"));
        }

        public Task<ApiResponse<List<EventResponse>>> GetEventsByPlaceAsync(Guid placeId, Guid? currentUserId = null)
        {
            return Task.FromResult(ApiResponse<List<EventResponse>>.ErrorResult("Event service not fully implemented yet"));
        }

        public async Task<ApiResponse<List<EventResponse>>> GetUpcomingEventsAsync(Guid userId, int days = 7)
        {
            try
            {
                var startDate = DateTime.UtcNow;
                var endDate = startDate.AddDays(days);

                var events = await _context.Events
                    .Include(e => e.Hobby)
                    .Include(e => e.Place)
                    .Include(e => e.CreatorUser)
                    .Include(e => e.Participants)
                    .Where(e => e.StartDateTime >= startDate && e.StartDateTime <= endDate)
                    .OrderBy(e => e.StartDateTime)
                    .Take(20)
                    .ToListAsync();

                var eventResponses = events.Select(e => new EventResponse
                {
                    EventId = e.EventId,
                    Title = e.Title,
                    Description = e.Description,
                    CreatorUserId = e.CreatorUserId,
                    CreatorName = e.CreatorUser?.Username ?? "Unknown",
                    HobbyId = e.HobbyId,
                    HobbyName = e.Hobby?.Name ?? "General",
                    PlaceId = e.PlaceId,
                    PlaceName = e.Place?.Name ?? e.CustomLocationName ?? "Custom Location",
                    Location = e.Place?.Address ?? e.CustomLocationAddress ?? "Location TBD",
                    Latitude = (double?)e.Latitude,
                    Longitude = (double?)e.Longitude,
                    StartTime = e.StartDateTime,
                    EndTime = e.EndDateTime,
                    MaxParticipants = e.MaxParticipants,
                    CurrentParticipants = e.Participants.Count,
                    Price = e.EntryFee,
                    ImageUrl = !string.IsNullOrEmpty(e.ImageUrls) ? e.ImageUrls : null,
                    Status = e.StartDateTime > DateTime.UtcNow ? "Upcoming" : "Live",
                    IsPublic = e.IsPublic,
                    RequiresApproval = e.RequiresApproval,
                    Requirements = null, // This property doesn't exist in the Event entity based on the error
                    CreatedAt = e.CreatedAt,
                    UpdatedAt = e.UpdatedAt,
                    IsParticipant = false, // This would need to be calculated based on current user
                    ParticipationStatus = null
                }).ToList();

                return ApiResponse<List<EventResponse>>.SuccessResult(eventResponses);
            }
            catch (Exception ex)
            {
                return ApiResponse<List<EventResponse>>.ErrorResult($"Failed to fetch upcoming events: {ex.Message}");
            }
        }

        public Task<ApiResponse<List<EventResponse>>> GetPopularEventsAsync(int count = 10, Guid? currentUserId = null)
        {
            return Task.FromResult(ApiResponse<List<EventResponse>>.ErrorResult("Event service not fully implemented yet"));
        }
    }
}








// using Microsoft.EntityFrameworkCore;
// using Same.Data;
// using Same.Models.DTOs.Responses;
// using Same.Services.Interfaces;

// namespace Same.Services.Implementations
// {
//     public class EventService : IEventService
//     {
//         private readonly ApplicationDbContext _context;

//         public EventService(ApplicationDbContext context)
//         {
//             _context = context;
//         }

//         public Task<ApiResponse<EventResponse>> CreateEventAsync(Guid creatorId, CreateEventRequest request)
//         {
//             return Task.FromResult(ApiResponse<EventResponse>.ErrorResult("Event service not fully implemented yet"));
//         }

//         public Task<ApiResponse<EventResponse>> GetEventByIdAsync(Guid eventId, Guid? currentUserId = null)
//         {
//             return Task.FromResult(ApiResponse<EventResponse>.ErrorResult("Event service not fully implemented yet"));
//         }

//         public Task<ApiResponse<List<EventResponse>>> GetEventsByCreatorAsync(Guid creatorId)
//         {
//             return Task.FromResult(ApiResponse<List<EventResponse>>.ErrorResult("Event service not fully implemented yet"));
//         }

//         public Task<ApiResponse<List<EventResponse>>> SearchEventsAsync(SearchEventsRequest request, Guid? currentUserId = null)
//         {
//             return Task.FromResult(ApiResponse<List<EventResponse>>.ErrorResult("Event service not fully implemented yet"));
//         }

//         public Task<ApiResponse<EventResponse>> UpdateEventAsync(Guid eventId, Guid creatorId, UpdateEventRequest request)
//         {
//             return Task.FromResult(ApiResponse<EventResponse>.ErrorResult("Event service not fully implemented yet"));
//         }

//         public Task<ApiResponse<bool>> DeleteEventAsync(Guid eventId, Guid creatorId)
//         {
//             return Task.FromResult(ApiResponse<bool>.ErrorResult("Event service not fully implemented yet"));
//         }

//         public Task<ApiResponse<bool>> JoinEventAsync(Guid eventId, Guid userId)
//         {
//             return Task.FromResult(ApiResponse<bool>.ErrorResult("Event service not fully implemented yet"));
//         }

//         public Task<ApiResponse<bool>> LeaveEventAsync(Guid eventId, Guid userId)
//         {
//             return Task.FromResult(ApiResponse<bool>.ErrorResult("Event service not fully implemented yet"));
//         }

//         public Task<ApiResponse<List<UserResponse>>> GetEventParticipantsAsync(Guid eventId)
//         {
//             return Task.FromResult(ApiResponse<List<UserResponse>>.ErrorResult("Event service not fully implemented yet"));
//         }

//         public Task<ApiResponse<List<EventResponse>>> GetUserEventsAsync(Guid userId)
//         {
//             return Task.FromResult(ApiResponse<List<EventResponse>>.ErrorResult("Event service not fully implemented yet"));
//         }

//         public Task<ApiResponse<bool>> InviteToEventAsync(Guid eventId, Guid creatorId, List<Guid> userIds)
//         {
//             return Task.FromResult(ApiResponse<bool>.ErrorResult("Event service not fully implemented yet"));
//         }

//         public Task<ApiResponse<List<EventResponse>>> GetNearbyEventsAsync(decimal latitude, decimal longitude, double radiusKm = 10, Guid? currentUserId = null)
//         {
//             return Task.FromResult(ApiResponse<List<EventResponse>>.ErrorResult("Event service not fully implemented yet"));
//         }

//         public Task<ApiResponse<List<EventResponse>>> GetEventsByHobbyAsync(Guid hobbyId, Guid? currentUserId = null)
//         {
//             return Task.FromResult(ApiResponse<List<EventResponse>>.ErrorResult("Event service not fully implemented yet"));
//         }

//         public Task<ApiResponse<List<EventResponse>>> GetEventsByPlaceAsync(Guid placeId, Guid? currentUserId = null)
//         {
//             return Task.FromResult(ApiResponse<List<EventResponse>>.ErrorResult("Event service not fully implemented yet"));
//         }

//         public async Task<ApiResponse<List<EventResponse>>> GetUpcomingEventsAsync(Guid userId, int days = 7)
//         {
//             try
//             {
//                 var startDate = DateTime.UtcNow;
//                 var endDate = startDate.AddDays(days);

//                 var events = await _context.Events
//                     .Include(e => e.Hobby)
//                     .Include(e => e.Place)
//                     .Include(e => e.Participants)
//                     .Where(e => e.StartDateTime >= startDate && e.StartDateTime <= endDate)
//                     .OrderBy(e => e.StartDateTime)
//                     .Take(20)
//                     .ToListAsync();

//                 var eventResponses = events.Select(e => new EventResponse
//                 {
//                     EventId = e.EventId,
//                     Title = e.Title,
//                     Description = e.Description,
//                     CreatorUserId = e.CreatorUserId,
//                     CreatorName = e.CreatorUser?.Username,
//                     HobbyId = e.HobbyId,
//                     HobbyName = e.Hobby?.Name ?? "General",
//                     PlaceId = e.PlaceId,
//                     PlaceName = e.Place?.Name,
//                     Location = e.Place?.Address ?? "Location TBD",
//                     Latitude = e.Place?.Latitude,
//                     Longitude = e.Place?.Longitude,
//                     StartTime = e.StartDateTime,
//                     EndTime = e.EndDateTime,
//                     MaxParticipants = e.MaxParticipants ?? 0,
//                     CurrentParticipants = e.Participants.Count,
//                     Price = e.Price,
//                     ImageUrl = e.ImageUrl,
//                     Status = e.StartDateTime > DateTime.UtcNow ? "Upcoming" : "Live",
//                     IsPublic = e.IsPublic,
//                     RequiresApproval = e.RequiresApproval,
//                     Requirements = e.Requirements,
//                     CreatedAt = e.CreatedAt,
//                     UpdatedAt = e.UpdatedAt,
//                     IsParticipant = false, // This would need to be calculated based on current user
//                     ParticipationStatus = null
//                 }).ToList();

//                 return ApiResponse<List<EventResponse>>.SuccessResult(eventResponses);
//             }
//             catch (Exception ex)
//             {
//                 return ApiResponse<List<EventResponse>>.ErrorResult($"Failed to fetch upcoming events: {ex.Message}");
//             }
//         }

//         public Task<ApiResponse<List<EventResponse>>> GetPopularEventsAsync(int count = 10, Guid? currentUserId = null)
//         {
//             return Task.FromResult(ApiResponse<List<EventResponse>>.ErrorResult("Event service not fully implemented yet"));
//         }
//     }
// }

File: hobby-hub-map/src/services/eventService.ts
// src/services/eventService.ts
import { apiClient } from './api';
import { UserResponse } from './authService';

export interface EventResponse {
  eventId: string;
  title: string;
  description?: string;
  creatorUserId: string;
  creatorName?: string;
  hobbyId: string;
  hobbyName?: string;
  placeId?: string;
  placeName?: string;
  location?: string;
  latitude?: number;
  longitude?: number;
  startTime: string;
  endTime?: string;
  maxParticipants: number;
  currentParticipants: number;
  price?: number;
  imageUrl?: string;
  status: string;
  isPublic: boolean;
  requiresApproval: boolean;
  requirements?: string;
  createdAt: string;
  updatedAt: string;
  isParticipant?: boolean;
  participationStatus?: string;
}

export interface CreateEventRequest {
  hobbyId: string;
  placeId?: string;
  title: string;
  description?: string;
  customLocationName?: string;
  customLocationAddress?: string;
  latitude?: number;
  longitude?: number;
  startDateTime: string;
  endDateTime?: string;
  maxParticipants?: number;
  entryFee?: number;
  requiresApproval?: boolean;
  isPublic?: boolean;
  imageUrls?: string[];
}

export interface UpdateEventRequest extends CreateEventRequest {
  isActive?: boolean;
  status?: string;
}

export interface SearchEventsRequest {
  title?: string;
  hobbyId?: string;
  placeId?: string;
  startDate?: string;
  endDate?: string;
  minPrice?: number;
  maxPrice?: number;
  latitude?: number;
  longitude?: number;
  radiusKm?: number;
  privacyLevel?: string;
  hasAvailableSlots?: boolean;
  page?: number;
  pageSize?: number;
}

export interface JoinEventRequest {
  eventId: string;
  message?: string;
  payEntryFee?: boolean;
}

class EventService {
  // Event retrieval
  async getUpcomingEvents(page: number = 1, pageSize: number = 20): Promise<EventResponse[]> {
    return await apiClient.get<EventResponse[]>(`/event?page=${page}&pageSize=${pageSize}`);
  }

  async getEventById(eventId: string): Promise<EventResponse> {
    return await apiClient.get<EventResponse>(`/event/${eventId}`);
  }

  async getEventsByHobby(hobbyId: string): Promise<EventResponse[]> {
    return await apiClient.get<EventResponse[]>(`/event/hobby/${hobbyId}`);
  }

  async getEventsByLocation(latitude: number, longitude: number, radiusKm: number = 10): Promise<EventResponse[]> {
    return await apiClient.get<EventResponse[]>(
      `/event/location?latitude=${latitude}&longitude=${longitude}&radiusKm=${radiusKm}`
    );
  }

  async getMyEvents(filter: string = 'all'): Promise<EventResponse[]> {
    return await apiClient.get<EventResponse[]>(`/event/my-events?filter=${filter}`);
  }

  async searchEvents(request: SearchEventsRequest): Promise<EventResponse[]> {
    return await apiClient.post<EventResponse[]>('/event/search', request);
  }

  async getPopularEvents(count: number = 10): Promise<EventResponse[]> {
    return await apiClient.get<EventResponse[]>(`/event/popular?count=${count}`);
  }

  // Event management
  async createEvent(request: CreateEventRequest): Promise<EventResponse> {
    return await apiClient.post<EventResponse>('/event', request);
  }

  async updateEvent(eventId: string, request: UpdateEventRequest): Promise<EventResponse> {
    return await apiClient.put<EventResponse>(`/event/${eventId}`, request);
  }

  async deleteEvent(eventId: string): Promise<boolean> {
    return await apiClient.delete<boolean>(`/event/${eventId}`);
  }

  // Event participation
  async joinEvent(eventId: string, request?: JoinEventRequest): Promise<boolean> {
    return await apiClient.post<boolean>(`/event/${eventId}/join`, request || {});
  }

  async leaveEvent(eventId: string): Promise<boolean> {
    return await apiClient.post<boolean>(`/event/${eventId}/leave`);
  }

  async getEventParticipants(eventId: string): Promise<UserResponse[]> {
    return await apiClient.get<UserResponse[]>(`/event/${eventId}/participants`);
  }

  async inviteToEvent(eventId: string, userIds: string[]): Promise<boolean> {
    return await apiClient.post<boolean>(`/event/${eventId}/invite`, { userIds });
  }

  // Utility methods for formatting
  formatEventDate(dateString: string): string {
    return new Date(dateString).toLocaleDateString('en-US', {
      weekday: 'long',
      year: 'numeric',
      month: 'long',
      day: 'numeric'
    });
  }

  formatEventTime(dateString: string): string {
    return new Date(dateString).toLocaleTimeString('en-US', {
      hour: '2-digit',
      minute: '2-digit'
    });
  }

  isEventUpcoming(dateString: string): boolean {
    return new Date(dateString) > new Date();
  }

  isEventLive(startString: string, endString?: string): boolean {
    const now = new Date();
    const start = new Date(startString);
    const end = endString ? new Date(endString) : null;
    
    return start <= now && (!end || now <= end);
  }

  getEventStatus(event: EventResponse): 'upcoming' | 'live' | 'ended' | 'cancelled' {
    if (event.status.toLowerCase() === 'cancelled') return 'cancelled';
    
    if (this.isEventLive(event.startTime, event.endTime)) return 'live';
    if (this.isEventUpcoming(event.startTime)) return 'upcoming';
    return 'ended';
  }
}

export const eventService = new EventService();












// import { apiClient } from './api';

// export interface Event {
//   eventId: string;
//   title: string;
//   description: string;
//   startDate: string;
//   endDate?: string;
//   location: {
//     latitude: number;
//     longitude: number;
//     address: string;
//   };
//   maxParticipants?: number;
//   currentParticipants: number;
//   hobbyId: string;
//   hobbyName: string;
//   createdBy: {
//     id: string;
//     username: string;
//     profileImage?: string;
//   };
//   isJoined?: boolean;
//   imageUrl?: string;
//   tags?: string[];
// }

// export interface EventResponse {
//   eventId: string;
//   title: string;
//   description: string;
//   startDate: string;
//   endDate?: string;
//   location: {
//     latitude: number;
//     longitude: number;
//     address: string;
//   };
//   maxParticipants?: number;
//   currentParticipants: number;
//   hobbyId: string;
//   hobbyName: string;
//   createdBy: {
//     id: string;
//     username: string;
//     profileImage?: string;
//   };
//   isJoined?: boolean;
//   imageUrl?: string;
//   tags?: string[];
// }

// export interface CreateEventRequest {
//   title: string;
//   description: string;
//   startDate: string;
//   endDate?: string;
//   location: {
//     latitude: number;
//     longitude: number;
//     address: string;
//   };
//   maxParticipants?: number;
//   hobbyId: string;
//   imageUrl?: string;
//   tags?: string[];
// }

// export interface JoinEventRequest {
//   message?: string;
// }

// class EventService {
//   async getUpcomingEvents(page: number = 1, pageSize: number = 20) {
//     return apiClient.get<EventResponse[]>(`/event?page=${page}&pageSize=${pageSize}`);
//   }

//   async getEventById(eventId: string) {
//     return apiClient.get<EventResponse>(`/event/${eventId}`);
//   }

//   async getEventsByHobby(hobbyId: string) {
//     return apiClient.get<EventResponse[]>(`/event/hobby/${hobbyId}`);
//   }

//   async getEventsByLocation(latitude: number, longitude: number, radiusKm: number = 10) {
//     return apiClient.get<EventResponse[]>(
//       `/event/location?latitude=${latitude}&longitude=${longitude}&radiusKm=${radiusKm}`
//     );
//   }

//   async createEvent(eventData: CreateEventRequest) {
//     return apiClient.post<EventResponse>('/event', eventData);
//   }

//   async joinEvent(eventId: string, joinData?: JoinEventRequest) {
//     return apiClient.post(`/event/${eventId}/join`, joinData);
//   }

//   async leaveEvent(eventId: string) {
//     return apiClient.post(`/event/${eventId}/leave`);
//   }

//   async getMyEvents(filter: 'all' | 'upcoming' | 'past' = 'all') {
//     return apiClient.get<EventResponse[]>(`/event/my-events?filter=${filter}`);
//   }
// }

// export const eventService = new EventService();
